keps/sig-architecture/4402-go-workspaces/README.md

```diff
@@ -0,0 +1,437 @@
+   1. Fix any remaining tools which had dependencies on the gengo-based tools.
+   1. Remove detritus, including (but not limited to) run-in-gopath.sh,
+      references to GOPATH, workarounds for vendored packages, etc.
+
+Along the way, some tools will need more touch than others.  At least one -
+import-boss - seems best to retool away from gengo entirely.
+
+Somewhere in there, we want to make deeper changes to gengo/v2 to simplify it.
+Some of the framework seemed clean at the time, but nearly a decade later, just
+makes the tools harder to understand.  This could happen after all of this
+work (repeating the sequence) or at the same time, making for larger but more
+complete reworking.
+
+### Alternatives to gengo/v2
+
+We did consider some options before landing on gengo/v2.
+
+  1. Put it in `gengo/v2` with a README that says "don't use this unless you
+     are part of kubernetes".  Pro: simple.
+  1. Put it in a new `k8s.io/<something>` repo with a README that says "don't
+     use this unless you are part of kubernetes".
+  1. Put it in `k8s.io/code-generator/<something>` with a README that says
+     "don't use this unless you are part of kubernetes" (pro: that puts it into
+     k/k so atomic commits are OK).
+  1. Try to make a k8s.io/internal github org work, and put it as
+     `k8s.io/internal/<something>`.  Pro: stronger than a README.
+
+The general feeling was that `gengo/v2` is simplest and "good enough".
+
+### Risks and Mitigations
+
+There's not a clean way to do this.  Go workspaces DOES NOT work with our
+symlinks-in-vendor hack, and all of our tools do not work without it.  We have
+a LOT of tools and the hacks run DEEP.  We have a decade of legacy to scrub
+off.  Worse, many of our tools make dubious assumptions about the
+interchangeability of Go package paths (e.g. "example.com/foo/bar") and disk
+paths.  Those assumptions held as long as we used GOPATH, BARELY held with
+staging modules, and break totally in workspaces.
+
+This means that some of our ecosystem-facing APIs (codegen tools, in
+particular) *must* change.  We can mitigate some of the impact by creating "v2"
+of things like k8s.io/gengo (which powers many tools).  We might choose to do
+the same in kube-openapi, though its API is not changing, just the
+implementation.
+
+However, k8s.io/code-generator is more challenging.  It holds all of our tools
+and some scripts for invoking them, and is used by a non-trivial number of
+downstream projects (operators, custom API servers, etc).  Making a v2 of this
+(which itself lives in staging, and is subject to this KEP!!) is something we
+have not done before.  The simple answer is to rely on the fact that it was
+never really versioned before, and just make the changes.  We could move it all
+to a new "legacy-code-generator" staging repo and explain that it gets no
+further support, or we could just tag the last legacy release of
+k8s.io/code-generator and tell people to sync to that tag.  Same effect.
```

Comment:
I'd rely on the existing version tags on code-generator ... folks with hard dependencies on the current CLI invocation can stick on the v0.29.x tags until they adapt to the invocation changes.